cwltool/process.py

```diff
@@ -226,27 +226,60 @@ def adjustFilesWithSecondary(rec, op, primary=None):
         for d in rec:
             adjustFilesWithSecondary(d, op, primary)
 
+def _stage_file(src, dst):
+    if src == dst:
+        return
+    if onWindows():
+        if os.path.isfile(src):
+            shutil.copy(src, dst)
+        elif os.path.isdir(src):
+            if os.path.exists(dst) and os.path.isdir(dst):
+                shutil.rmtree(dst)
+            copytree_with_merge(src, dst)
+    else:
+        os.symlink(src, dst)
+
+def _stage_file_with(action, source_directories, fs_access):
+    def _stage(src, dst):
+        if src == dst:
+            return
+
+        # If the source is not contained in source_directories we're not allowed to delete it
+        src_can_deleted = any(os.path.commonprefix([p, src]) == p for p in source_directories)
+
+        _action = "move" if action == "move" and src_can_deleted else "copy"
+
+        if _action == "move":
+            _logger.debug("Moving %s to %s", src, dst)
+            if fs_access.isdir(src) and fs_access.isdir(dst):
+                # merge directories
+                for dir_entry in scandir(src):
+                    _stage(dir_entry.path, fs_access.join(dst, dir_entry.name))
+            else:
+                shutil.move(src, dst)
 
-def stageFiles(pm, stageFunc=None, ignoreWritable=False, symLink=True, secret_store=None):
+        elif _action == "copy":
+            _logger.debug("Copying %s to %s", src, dst)
+            if fs_access.isdir(src):
+                if os.path.isdir(dst):
+                    shutil.rmtree(dst)
+                elif os.path.isfile(dst):
+                    os.unlink(dst)
+                shutil.copytree(src, dst)
+            else:
+                shutil.copy2(src, dst)
+    return _stage
+
+
+def stageFiles(pm, stageFunc=_stage_file, ignoreWritable=False, secret_store=None):
     # type: (PathMapper, Callable[..., Any], bool, bool, SecretStore) -> None
```

Comment:
`error: Type signature has too many arguments`
```suggestion
    # type: (PathMapper, Callable[..., Any], bool, SecretStore) -> None
```